[libc++] Speed up classic locale

dvyukov opened the pull request:
Locale objects use atomic reference counting, which may be very expensive in parallel applications. The classic locale is used by default by all streams and can be very contended. But it's never destroyed, so the reference counting is also completely pointless on the classic locale. Currently ~70% of time in the parallel stringstream benchmarks is spent in locale ctor/dtor. And the execution radically slows down with more threads.

Avoid reference counting on the classic locale and inline common ctors/dtor. With this change locale ctor/dtor time become negligible and the benchmark starts to scale with threads.

```
                                       │ bench/stream.baseline.res │     bench/stream.optimized.res      │
                                       │          sec/op           │   sec/op     vs base                │
Istream_numbers/0/real_time/threads:1                  4.686µ ± 0%   4.213µ ± 0%  -10.09% (p=0.000 n=24)
Istream_numbers/0/real_time/threads:72               339.712µ ± 4%   8.939µ ± 2%  -97.37% (p=0.000 n=24)
Istream_numbers/1/real_time/threads:1                  4.900µ ± 0%   4.773µ ± 0%   -2.60% (p=0.000 n=24)
Istream_numbers/1/real_time/threads:72                49.308µ ± 3%   9.694µ ± 1%  -80.34% (p=0.000 n=24)
Istream_numbers/2/real_time/threads:1                  4.902µ ± 0%   4.769µ ± 0%   -2.72% (p=0.000 n=24)
Istream_numbers/2/real_time/threads:72                 416.6µ ± 9%   410.5µ ± 6%        ~ (p=0.111 n=24)
Istream_numbers/3/real_time/threads:1                  4.716µ ± 0%   4.737µ ± 0%   +0.45% (p=0.000 n=24)
Istream_numbers/3/real_time/threads:72                 419.7µ ± 6%   458.3µ ± 4%   +9.19% (p=0.003 n=24)
Ostream_number/0/real_time/threads:1                   184.0n ± 1%   135.0n ± 1%  -26.63% (p=0.000 n=24)
Ostream_number/0/real_time/threads:72                17279.0n ± 4%   308.0n ± 1%  -98.22% (p=0.000 n=24)
Ostream_number/1/real_time/threads:1                   253.5n ± 1%   199.0n ± 1%  -21.50% (p=0.000 n=24)
Ostream_number/1/real_time/threads:72                13957.5n ± 2%   382.0n ± 0%  -97.26% (p=0.000 n=24)
Ostream_number/2/real_time/threads:1                   253.0n ± 1%   202.0n ± 1%  -20.16% (p=0.000 n=24)
Ostream_number/2/real_time/threads:72                  28.04µ ± 4%   18.80µ ± 7%  -32.96% (p=0.000 n=24)
Ostream_number/3/real_time/threads:1                   187.0n ± 1%   190.5n ± 0%   +1.87% (p=0.000 n=24)
Ostream_number/3/real_time/threads:72                  20.25µ ± 7%   20.37µ ± 5%        ~ (p=0.736 n=24)

```

dvyukov committed:
[libc++] Speed up classic locale

Locale objects use atomic reference counting, which may be very expensive
in parallel applications. The classic locale is used by default by all
streams and can be very contended. But it's never destroyed, so the
reference counting is also completely pointless on the classic locale.
Currently ~70% of time in the parallel stringstream benchmarks is spent
in locale ctor/dtor. And the execution radically slows down with more threads.

Avoid reference counting on the classic locale and inline common ctors/dtor.
With this change locale ctor/dtor time become negligible and the benchmark
starts to scale with threads.

                               │baseline sec/op│ optimized sec/op                  │
Ostream_number/threads:1           184.5n ± 0%   133.0n ± 1%  -27.91% (p=0.000 n=30)
Ostream_number/threads:72        24188.0n ± 3%   321.0n ± 2%  -98.67% (p=0.000 n=30)
Istream_numbers/1024/threads:1     4.667µ ± 1%   4.273µ ± 0%   -8.43% (p=0.000 n=30)
Istream_numbers/1024/threads:72  559.657µ ± 1%   9.350µ ± 1%  -98.33% (p=0.000 n=30)

philnik777 commented:
The numbers look really good, but as-is this is almost certainly ABI-breaking. You can't just move functions to and from the dylib. From taking a quick look at the changes, I think this can be done without breaking the ABI. I'd also like to see benchmarks for other locales, since they will probably be affected negatively by this change.

firewave commented:
See #56202 for a related issue which also contains some benchmark code which might be covering what @philnik777 requested .

dvyukov commented:
> I think this can be done without breaking the ABI

How? Do you have any specific way in mind? Are there examples of such changes?

dvyukov commented:
> See https://github.com/llvm/llvm-project/issues/56202 for a related issue which also contains some benchmark code which might be covering what @philnik777 requested .

I don't see any benchmarks there, nor in referenced changes. Which ones do you mean?

dvyukov committed:
add _LIBCPP_HIDE_FROM_ABI to inline functions

dvyukov committed:
extend benchmarks to non-classic locale

                                       │ bench/stream.baseline.res │     bench/stream.optimized.res      │
                                       │          sec/op           │   sec/op     vs base                │
Istream_numbers/0/real_time/threads:1                  4.686µ ± 0%   4.213µ ± 0%  -10.09% (p=0.000 n=24)
Istream_numbers/0/real_time/threads:72               339.712µ ± 4%   8.939µ ± 2%  -97.37% (p=0.000 n=24)
Istream_numbers/1/real_time/threads:1                  4.900µ ± 0%   4.773µ ± 0%   -2.60% (p=0.000 n=24)
Istream_numbers/1/real_time/threads:72                49.308µ ± 3%   9.694µ ± 1%  -80.34% (p=0.000 n=24)
Istream_numbers/2/real_time/threads:1                  4.902µ ± 0%   4.769µ ± 0%   -2.72% (p=0.000 n=24)
Istream_numbers/2/real_time/threads:72                 416.6µ ± 9%   410.5µ ± 6%        ~ (p=0.111 n=24)
Istream_numbers/3/real_time/threads:1                  4.716µ ± 0%   4.737µ ± 0%   +0.45% (p=0.000 n=24)
Istream_numbers/3/real_time/threads:72                 419.7µ ± 6%   458.3µ ± 4%   +9.19% (p=0.003 n=24)
Ostream_number/0/real_time/threads:1                   184.0n ± 1%   135.0n ± 1%  -26.63% (p=0.000 n=24)
Ostream_number/0/real_time/threads:72                17279.0n ± 4%   308.0n ± 1%  -98.22% (p=0.000 n=24)
Ostream_number/1/real_time/threads:1                   253.5n ± 1%   199.0n ± 1%  -21.50% (p=0.000 n=24)
Ostream_number/1/real_time/threads:72                13957.5n ± 2%   382.0n ± 0%  -97.26% (p=0.000 n=24)
Ostream_number/2/real_time/threads:1                   253.0n ± 1%   202.0n ± 1%  -20.16% (p=0.000 n=24)
Ostream_number/2/real_time/threads:72                  28.04µ ± 4%   18.80µ ± 7%  -32.96% (p=0.000 n=24)
Ostream_number/3/real_time/threads:1                   187.0n ± 1%   190.5n ± 0%   +1.87% (p=0.000 n=24)
Ostream_number/3/real_time/threads:72                  20.25µ ± 7%   20.37µ ± 5%        ~ (p=0.736 n=24)

firewave commented:
> > See #56202 for a related issue which also contains some benchmark code which might be covering what @philnik777 requested .
> 
> I don't see any benchmarks there, nor in referenced changes. Which ones do you mean?

https://github.com/llvm/llvm-project/issues/56202#issuecomment-1166229891

dvyukov commented:
I've extended benchmarks to test non-classic locale, and update benchmark results in the PR description.

dvyukov commented:
I've added _LIBCPP_HIDE_FROM_ABI to inline functions as the bot suggested.

Now some tests are failing with dynamic libc++.so, e.g.:
```
# | t.tmp.exe: libcxx/test/std/localization/locales/locale/locale.cons/default.pass.cpp:81: int main(int, char **): Assertion `loc2 == std::locale(LOCALE_en_US_UTF_8)' failed.
```

Is it possible to implement new inline functions in libc++?

ldionne commented on libcxx/benchmarks/stringstream.bench.cpp:
```diff
@@ -14,17 +15,72 @@ double istream_numbers() {
+    static std::mutex mu;
+    std::lock_guard l(mu);
+    switch (state.range(0)) {
+    case 0: {
```
Is there any reason not to prefer having 4 different benchmarks instead?

dvyukov replied:
It's all different benchamarks:
```
Ostream_number/0/real_time/threads:1                   184.0n ± 1%   135.0n ± 1%  -26.63% (p=0.000 n=24)
Ostream_number/0/real_time/threads:72                17279.0n ± 4%   308.0n ± 1%  -98.22% (p=0.000 n=24)
Ostream_number/1/real_time/threads:1                   253.5n ± 1%   199.0n ± 1%  -21.50% (p=0.000 n=24)
Ostream_number/1/real_time/threads:72                13957.5n ± 2%   382.0n ± 0%  -97.26% (p=0.000 n=24)
Ostream_number/2/real_time/threads:1                   253.0n ± 1%   202.0n ± 1%  -20.16% (p=0.000 n=24)
Ostream_number/2/real_time/threads:72                  28.04µ ± 4%   18.80µ ± 7%  -32.96% (p=0.000 n=24)
Ostream_number/3/real_time/threads:1                   187.0n ± 1%   190.5n ± 0%   +1.87% (p=0.000 n=24)
Ostream_number/3/real_time/threads:72                  20.25µ ± 7%   20.37µ ± 5%        ~ (p=0.736 n=24)
```

Or you mean copy-pasting and specializing the top benchmark function for each case?
Do you think it will be better? I would assume it will increase code size and add some duplication.

ldionne reviewed (changes requested):
As Nikolas mentioned, this is an ABI break. The speedup is interesting, but we'd need to either find a way to make this not an ABI break, or to only enable the optimization under an ABI flag.

Is there any reason why we're not storing the classic locale pointer you're adding to the `locale` class inside the `__imp` class instead? I think that would allow not breaking the ABI, since that's a pimpl idiom.

dvyukov commented:
> As Nikolas mentioned, this is an ABI break. The speedup is interesting, but we'd need to either find a way to make this not an ABI break, or to only enable the optimization under an ABI flag.
> 
> Is there any reason why we're not storing the classic locale pointer you're adding to the `locale` class inside the `__imp` class instead? I think that would allow not breaking the ABI, since that's a pimpl idiom.

I don't understand the problem space well, so I am open to suggestions re ABI.
What exactly is breaking ABI here? New methods? Inline methods? The static field?
What is ABI flag? How bad is this?
What are other options? Were there similar precendents?

Re __imp/pimpl, the original motivation was inlining. __imp is not in the header, so we can get access to it from any inline methods.
If there are no good options that preserve inlining, I guess we can give up on it. It contributes to this speed somewhat:
```
Ostream_number/0/real_time/threads:1                   184.0n ± 1%   135.0n ± 1%  -26.63% (p=0.000 n=24)
```

But for this, more important one, it's probably in the noise:
```
Ostream_number/0/real_time/threads:72                17279.0n ± 4%   308.0n ± 1%  -98.22% (p=0.000 n=24)
```

philnik777 commented:
> > As Nikolas mentioned, this is an ABI break. The speedup is interesting, but we'd need to either find a way to make this not an ABI break, or to only enable the optimization under an ABI flag.
> > Is there any reason why we're not storing the classic locale pointer you're adding to the `locale` class inside the `__imp` class instead? I think that would allow not breaking the ABI, since that's a pimpl idiom.
> 
> I don't understand the problem space well, so I am open to suggestions re ABI. What exactly is breaking ABI here? New methods? Inline methods? The static field?

The ABI break is actually that you remove functions from the dylib. Adding new ones isn't a huge problem.

> What is ABI flag? How bad is this? What are other options? Were there similar precedents?

An ABI flag is a macro that is only defined if it's OK to break the ABI. They are listed [here](https://github.com/llvm/llvm-project/blob/e734bc53de8173a1b54493f3e0c8f70575341a9e/libcxx/include/__config#L98-L171).

> Re __imp/pimpl, the original motivation was inlining. __imp is not in the header, so we can get access to it from any inline methods. If there are no good options that preserve inlining, I guess we can give up on it. It contributes to this speed somewhat:
> 
> ```
> Ostream_number/0/real_time/threads:1                   184.0n ± 1%   135.0n ± 1%  -26.63% (p=0.000 n=24)
> ```
> 
> But for this, more important one, it's probably in the noise:
> 
> ```
> Ostream_number/0/real_time/threads:72                17279.0n ± 4%   308.0n ± 1%  -98.22% (p=0.000 n=24)
> ```

I'm a bit confused about this. The statements here seem contradictory. Could you try to reword this?

dvyukov commented:
> I'm a bit confused about this. The statements here seem contradictory. Could you try to reword this?

This change does 2 optimizations:
1. Allow inlining of the common locale ctor/dtor.
2. Avoid atomic reference counting for the classic locale.

(1) has small fixed-size impact (shaves off few cycles).
(2) may have very profound effect on highly threaded apps that use locale/streams concurrently (these numbers like -98.22%).

If inlining creates problems that are hard to resolve, then I am ready to drop it from this change and leave only optimization (2).

philnik777 commented:
> > I'm a bit confused about this. The statements here seem contradictory. Could you try to reword this?
> 
> This change does 2 optimizations:
> 
>     1. Allow inlining of the common locale ctor/dtor.
> 
>     2. Avoid atomic reference counting for the classic locale.
> 
> 
> (1) has small fixed-size impact (shaves off few cycles). (2) may have very profound effect on highly threaded apps that use locale/streams concurrently (these numbers like -98.22%).
> 
> If inlining creates problems that are hard to resolve, then I am ready to drop it from this change and leave only optimization (2).

Ah, OK. I think that is the way to go right now. That should simplify this patch quite a bit. We can still look into inlining the functions later if we want to. Dropping this part will most likely also resolve any ABI issues.

dvyukov commented:
> The ABI break is actually that you remove functions from the dylib. Adding new ones isn't a huge problem.

Ah, I see.

The problem is removing from dylib methods that are called by user code (ctor/dtor).
Removing internal methods (like locale::__imp::make_global()) is not a problem, because they are not called by user code.
Right?

If I try to resolve this, is there a way to test this?
IIRC ninja check-cxx passed for me.

Is we take an address of the ctor and assign to some global volatile var in the source file, then it should force the ctor symbol into dylib.
However, this is not possible for dtor, since it's not possible to take dtor address (at least I did not find any).

philnik777 commented:
> > The ABI break is actually that you remove functions from the dylib. Adding new ones isn't a huge problem.
> 
> Ah, I see.
> 
> The problem is removing from dylib methods that are called by user code (ctor/dtor). Removing internal methods (like locale::__imp::make_global()) is not a problem, because they are not called by user code. Right?

Yes, exactly.

> If I try to resolve this, is there a way to test this? IIRC ninja check-cxx passed for me.
> 
> Is we take an address of the ctor and assign to some global volatile var in the source file, then it should force the ctor symbol into dylib. However, this is not possible for dtor, since it's not possible to take dtor address (at least I did not find any).

We have `check-cxx-abilist` for this. It checks that the symbols exported from the dylib match the ones listed in the `abilist` files in `libcxx/lib/abi/`.

dvyukov commented:
> We have check-cxx-abilist for this. It checks that the symbols exported from the dylib match the ones listed in the abilist files in libcxx/lib/abi/.

Do I need to configure with some special flags for this?

```
ninja check-cxxabi-abilist
ninja: error: unknown target 'check-cxxabi-abilist'
```

I do have `check-cxxabi` and it passes. I configured with:

```
CC="clang -mavx2" CXX="clang++ -mavx2" cmake -DCMAKE_BUILD_TYPE=Release -DLLVM_TARGETS_TO_BUILD=X86 -DLLVM_ENABLE_PROJECTS="clang;clang-tools-extra;compiler-rt;lld;libc" -DLLVM_ENABLE_RUNTIMES="libcxx;libcxxabi" -DLIBC_INCLUDE_BENCHMARKS=Yes -GNinja -DLIBCXX_INCLUDE_BENCHMARKS=YES -DLIBCXX_ENABLE_STATIC_ABI_LIBRARY=ON -DLIBCXX_ENABLE_STATIC=YES ../llvm
```

dvyukov commented:
It was supposed to be:
```
ninja check-cxx-abilist
ninja: error: unknown target 'check-cxx-abilist'
```

I tried `check-cxxabi-abilist` as well just in case.

dvyukov commented:
I've sent separate #72112 without inlining, please take a look.
Yes, it's much cleaner and simpler, and keeps most of the benefit.

ldionne commented:
Thanks, I think #72112 is likely to make more progress since its scope is more narrow yet we reap most of the benefits. I'll close this for now and will review #72112 instead.
